src/txgraph.cpp

```diff
@@ -2322,13 +2328,22 @@ std::optional<std::pair<std::vector<TxGraph::Ref*>, FeePerWeight>> BlockBuilderI
         const auto& chunk_data = *m_cur_iter;
         const auto& chunk_end_entry = m_graph->m_entries[chunk_data.m_graph_index];
         auto cluster = chunk_end_entry.m_locator[0].cluster;
-        ret->first.resize(chunk_data.m_chunk_count);
-        auto start_pos = chunk_end_entry.m_main_lin_index + 1 - chunk_data.m_chunk_count;
-        bool is_end = cluster->GetClusterRefs(*m_graph, ret->first, start_pos);
-        if (is_end) {
+        if (chunk_data.m_chunk_count == LinearizationIndex(-1)) {
+            // Special case in case just a single transaction remains, avoiding the need to
+            // dispatch to and dereference Cluster.
+            ret->first.resize(1);
+            Assume(chunk_end_entry.m_ref != nullptr);
+            ret->first[0] = chunk_end_entry.m_ref;
             m_cur_cluster = nullptr;
         } else {
-            Assume(cluster == m_cur_cluster);
+            ret->first.resize(chunk_data.m_chunk_count);
+            auto start_pos = chunk_end_entry.m_main_lin_index + 1 - chunk_data.m_chunk_count;
+            bool is_end = cluster->GetClusterRefs(*m_graph, ret->first, start_pos);
+            if (is_end) {
+                m_cur_cluster = nullptr;
```

Comment:
could `Assume()` here that `chunk_data.m_chunk_count > 1`, since it would have been given the special value otherwise?